mpHealth-4.0 file-based health checks

Channyboy opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

################################################################################################


This is for #30056 .
This PR covers the introduction of the file-based health checking mechanism along with its configuration.
Functionality is beta-guarded.

- Introduces the `io.openliberty.microprofile.4.0.internal`  to incorporate the file-based functionality. Based on previous mpHealth bundles, had to extend the interface and impl hierarchies.
- Changed `mpHealth-4.0` feature to start the  `io.openliberty.microprofile.4.0.internal` bundle.
- Introduces `AppTracker40Impl` as we need to have controlover application started event to start the file-based health check processes
- Along with `AppTracker40Impl`, `HealthCheckService40Impl`, `FileHealthCheckBuilder` and `HealthFileUtils` are classes associated to the file-based health check functionality

Channyboy committed:
Initial; create separate bundle for mpHealth-4.0 for clear separation of new functionality

Channyboy committed:
Add new AppTracker40Impl service, needed for specific logic for file based health checks

Channyboy committed:
Refactor the HealthCheck40ServiceImpl health check method and introduced FileHealthCheck class

Channyboy committed:
Refactored out File I/O to utility class, Initial check for started/ready/live, implemented TimerTasks with static interval, cancel started process if file created, Using the FileHealthCheckBuilder similar to HealthCheckHTTPResponse builder

Channyboy committed:
For the first time, introduce configuration element for mpHealth w/ the fileUpdateInterval element and implement config parsing in the mpHealth-4.0 code for this

Channyboy committed:
Implemented logging with Tr, java doc , comments, refactored method names, removed unused code.

- Removed unused code in AppTracker40Impl, added comments.
- Added comments/logging to FileHealthCheckBuilder.java. Slight refactor of logic, prev lastModified was issued after a file was crated.
- Added Javadoc to HealthCheck40SErvice.java (Interface)
- Removed synchronization from get*File() in HealthFileUtils.java, it'll only be ever called once; Added comments, Javadoc, logging
- Added logging, Javadoc, comments, removed sysouts, refactored method names to HealthCheck40SErviceImpl

Channyboy committed:
Dynamic update for the fileUpdateInterval config

Channyboy committed:
Various items:

- FileHealthCheckBuilder moved from inner class to its own class
- FileUtils renamed HealthFileUtils.
- method for checking write perms now uses the strategy to create temp
  file
- Method for validating sytem renamed to initHealthFileValidation and
  returns boolean and is moved to HealthFileUtils
- Update HealthSerice40Service interface to incldue method to start
  fileHealthCheck process and new method to performFileHealthCheck
- TimerTask (i.e. FileUpdateProcess.java) for file health checking is now inner class of
  HealthCheck40SErviceImpl
- Removed FileHealthCheck class
- The mechanism for file HealthCheck prcoess is now that HS40Impl when
  activated will validate system for file health checking compataibility
  (i.e. IO perm) and then when
  the first application is started it will kick off the process through
  AppTracker40Impl
- Need to reintroduce the original.bnd
- Need to check MP Config each time (to satisfy checkpoint/restore condition)

Channyboy commented:
#build

Channyboy commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=1778e83b-2ed9-433d-9050-3880b061ce4c) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy commented:
Errors/Warnings encountered in above build are related to zos ejb and/or gui. Not a result of contents of this PR.

Channyboy commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=51cab84e-8098-4611-8e7b-80ccc03bd78d) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/internal/AppTracker40Impl.java:
```diff
@@ -0,0 +1,85 @@
+                    Tr.debug(tc, "applicationStarted(): started app updated in appStateMap = " + appStateMap.toString() + " for app: " + appName);
+                }
+            }
+            if (ProductInfo.getBetaEdition()) {
```
Shouldn't the beta guard be at the start of the method, so the extra replacing of the appState isn't done, in a non-beta release?

Channyboy replied:
That's the original logic from `AppTrackerImpl`. The method is overridden so the new code could be added, but would need to _reimplement_ the old logic because of the override.

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/internal/FileHealthCheckBuilder.java:
```diff
@@ -0,0 +1,102 @@
+     * Never used.
+     * Ensure the other constructor with File param is used.
+     */
+    private FileHealthCheckBuilder() {
```
Do we need this, if its never used?

Channyboy replied:
It's to ensure nobody would ever create `FileHealthCheckBuilder` suing the default constructor.

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/internal/HealthCheck40ServiceImpl.java:
```diff
@@ -0,0 +1,594 @@
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(tc, "HealthCheckServiceImpl is deactivated");
+        }
+
```
Should we call `stopTimers()` here too, just to be safe?

Channyboy replied:
Sure, we can.

Channyboy replied:
fixed with https://github.com/OpenLiberty/open-liberty/pull/30811/commits/96f6b7f3a6322a9f5290bc471b75ec0c4028c4dc

Channyboy commented:
Previous build stuck on `releasing open liberty code` Issuing new build.
#build  ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6b3f3844-4789-4ccd-a7a5-c536d9306f5c) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/internal/HealthCheck40ServiceImpl.java:
```diff
@@ -0,0 +1,594 @@
+    /**
+     * Resolve MP Config properties at startup and set default status.
+     */
+    private void resolveDefaultStatues() {
```
Typo: statues -> statuses

Channyboy replied:
fixed with https://github.com/OpenLiberty/open-liberty/pull/30811/commits/96f6b7f3a6322a9f5290bc471b75ec0c4028c4dc

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/internal/HealthFileUtils.java:
```diff
@@ -0,0 +1,251 @@
+        if (!healthDirFile.exists()) {
+            if (!HealthFileUtils.createDirectory(healthDirFile)) {
+                //TODO: Create the real message ID
+                Tr.warning(tc, "Unable to create the /health directory. The mpHealth file-based health check reporting will be disabled.");
```
Should all the warning msgs, be debug msgs for beta, and then create new messages with messageIDs later?

Channyboy replied:
fixed with https://github.com/OpenLiberty/open-liberty/pull/30811/commits/96f6b7f3a6322a9f5290bc471b75ec0c4028c4dc

pgunapal commented on dev/io.openliberty.microprofile.health.4.0.internal/src/io/openliberty/microprofile/health40/internal/HealthFileUtils.java:
```diff
@@ -0,0 +1,251 @@
+    static boolean canWriteToDirectory(File parentDir) {
+        Object token = ThreadIdentityManager.runAsServer();
+        try {
+
```
new line...

Channyboy replied:
fixed with https://github.com/OpenLiberty/open-liberty/pull/30811/commits/96f6b7f3a6322a9f5290bc471b75ec0c4028c4dc

pgunapal commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,22 @@
+mpHealth.desc=Configuration for MicroProfile Health.
+
+fileUpdateInterval=Interval time in for updating file based health checks.
+fileUpdateInterval.desc=Values specified will set the period interval for which the file based health checks are updated.
```
Should we mention that only seconds and milliseconds are supported?

Channyboy replied:
fixed with https://github.com/OpenLiberty/open-liberty/pull/30811/commits/96f6b7f3a6322a9f5290bc471b75ec0c4028c4dc

pgunapal commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -0,0 +1,29 @@
+    <OCD id="io.openliberty.microprofile.health" name="%mpHealth" description="%mpHealth.desc" ibm:alias="mpHealth" ibm:beta="true">
+        <!--  filePeriodSecond defaults to 10 seconds, but we do not specify default here because we'll programmatically do it. 
+        So we can throw warning if they set this, but havn't enabled file checking -->
+        <AD id="fileUpdateInterval" name="%fileUpdateInterval" description="%fileUpdateInterval.desc" required="false" type="String" ibm:beta="true"/>
```
Can't you add the `ibm:type="duration(ms)" default="10s"` to this, for better handling of error scenarios, instead of programmatically re-inventing the logic?

Channyboy replied:
The presence of the configuration attribute enables the functionality.
A default value can not be discernable from a user provided 10s.

pgunapal replied:
Won't the metatype server config, handle the cases, for user errors, if they typed something like "abcs", it would automatically throw an error, instead of us explicitly handling that? Since, we set the meta type to `ibm:type="duration(ms)"?

Channyboy replied:
Topic worth considering for a follow up PR?
Its recommended to use only "eng" time units now for  portability reasons., but ibm:type  already has pre translated units which are _all_ accepted
.

The sensible units for this config is `ms` and `s`.
But if we use `ibm:type="duration(s)"`  if the user typed `9z` by accident, that indicates 9 days for the interval and the config will accept it.

pgunapal reviewed (changes requested):
Looks good, minor comments. It would be good to get the messages from the metatype.properties reviewed by ID/L2.

Channyboy committed:
Various items:

- Change mpHealth-4.0 feature to point to the  new health.4.0.internal bundle (as well as the req to load the common bundle).
- File-based HealthCheck functionality is betaguarded
- mpHealth Config element and attribute is betaguarded
- misc fixes: warning was blocked by debug trace spec, copyright update, rename test to match version level
- Start process if no apps present; do an explicit start health check when last app has started
- Updated description for config values

Channyboy committed:
Address PR comments; another explicity stopTimers() for deactivate, fix typo for method name and use debug in place of warnings with no message IDs

Channyboy commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=0b680da0-4009-44cc-85e9-e581c07a4f3c) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

Channyboy commented:
@OpenLiberty/id-approvers 
You guys weren't tagged automatically, but does `[dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties](https://github.com/OpenLiberty/open-liberty/pull/30811/files#diff-3c8f1e97f55b5718b483fcb336576a43d55dfa38d52531ee0f453c2ea6cca2e3) need to be ID reviewed?

pgunapal reviewed (dismissed):
LGTM

dmuelle commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,22 @@
+mpHealth=MicroProfile Health
+mpHealth.desc=Configuration for MicroProfile Health.
+
+fileUpdateInterval=Interval time for updating the file based health checks.
```
```suggestion
fileUpdateInterval=Interval time at which to update the file-based health checks.
```

dmuelle replied:
what is getting "update" in tis context? Are you updating a file or the check? Or running a check and updating a file?

Channyboy replied:
It runs a check at intervals and updates a file if the checks are successful. The update consists of it _touching the file_ (i.e. modifies last accessed time). Hopefully the description entry explains it better. And it would appear its the description entry that is published on the reference documentation, and this value isn't used (but needed for

dmuelle commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,22 @@
+mpHealth.desc=Configuration for MicroProfile Health.
+
+fileUpdateInterval=Interval time for updating the file based health checks.
+fileUpdateInterval.desc=Value specified will set the interval for when the health checks are queried for the file based health checks. The value is a number followed by an optional time unit of "ms" (for milliseconds) or "s" (for seconds). If no time unit is specified then it is implied the value will be in seconds.
```
```suggestion
fileUpdateInterval.desc=The value that you specify sets the interval  at which the health checks are queried for the file-based health checks. The value is a number followed by an optional time unit of ms for milliseconds or s for seconds. If no time unit is specified, the value is in seconds by default.
```

dmuelle reviewed (changes requested):
Id review

Channyboy committed:
Apply ID  suggestions from code review

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

Channyboy commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=946596b1-1381-463d-8f16-ba530415ffed) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

dmuelle commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,22 @@
+mpHealth=MicroProfile Health
+mpHealth.desc=Configuration for MicroProfile Health.
+
+fileUpdateInterval=Interval time at which to update the file-based health checks.
```
```suggestion
fileUpdateInterval=Interval of time at which to update the file-based health checks.
```

dmuelle commented on dev/io.openliberty.microprofile.health.internal.common/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,22 @@
+mpHealth.desc=Configuration for MicroProfile Health.
+
+fileUpdateInterval=Interval time at which to update the file-based health checks.
+fileUpdateInterval.desc=The value that you specify sets the interval  at which the health checks are queried for the file-based health checks. The value is a number followed by an optional time unit of ms for milliseconds or s for seconds. If no time unit is specified, the value is in seconds by default.
```
```suggestion
fileUpdateInterval.desc=The value that you specify sets the interval at which the health checks are queried for the file-based health checks. The value is a number followed by an optional time unit of ms for milliseconds or s for seconds. If no time unit is specified, the value is in seconds by default.
```

dmuelle reviewed (changes requested):
a couple more minor edits-sorry think I introduced an extra space in my previous suggestion

Channyboy committed:
Apply ID suggestions from code review

Further ID change suggestions.

Co-authored-by: David Mueller <48686014+dmuelle@users.noreply.github.com>

Channyboy commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=8aa91723-54f5-41c7-9180-c74180cf3e9a) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._
